applications/CoSimulationApplication/co_simulation_api/co_sim_io_impl.h

```diff
@@ -0,0 +1,229 @@
+//        | |   / _ \___ \| | '_ ` _ \| | | | |/ _` | __| |/ _ \| '_  |
+//        | |__| (_) |__) | | | | | | | |_| | | (_| | |_| | (_) | | | |
+//         \____\___/____/|_|_| |_| |_|\__,_|_|\__,_|\__|_|\___/|_| |_|
+//
+//  License:		 BSD License
+//                   license: CoSimulationApplication/license.txt
+//
+//  Main authors:    Philipp Bucher
+//
+
+#ifndef KRATOS_CO_SIM_IO_IMPL_H_INCLUDED
+#define KRATOS_CO_SIM_IO_IMPL_H_INCLUDED
+
+// Optional includes
+#ifdef KRATOS_CO_SIM_IO_ENABLE_SOCKETS
+#include "co_sim_sockets_comm.h"
+#endif /* KRATOS_CO_SIM_IO_ENABLE_SOCKETS */
+
+
+#ifdef KRATOS_CO_SIM_IO_ENABLE_MPI
+#include "co_sim_mpi_comm.h"
+#endif /* KRATOS_CO_SIM_IO_ENABLE_MPI */
+
+// System includes
+#include <iostream>
+#include <sstream>
+#include <fstream>
+#include <vector>
+#include <string>
+#include <stdexcept>
+
+// Project includes
+#include "co_sim_file_comm.h"
+
+
+namespace CoSim {
+
+inline CoSimIO::CoSimIO(const std::string& rName, const std::string& rSettingsFileName, const bool IsConnectionMaster)
+    : CoSimIO::CoSimIO(rName, Internals::ReadSettingsFile(rSettingsFileName), IsConnectionMaster) { } // forwarding constructor call
+
+inline CoSimIO::CoSimIO(const std::string& rName, SettingsType rSettings, const bool IsConnectionMaster)
+{
+    Initialize(rName, rSettings, IsConnectionMaster);
+}
+
+inline bool CoSimIO::Connect()
+{
+    return mpComm->Connect();
+}
+
+inline bool CoSimIO::Disconnect()
+{
+    return mpComm->Disconnect();
+}
+
+
+inline void CoSimIO::SendControlSignal(const Internals::ControlSignal Signal, const std::string& rIdentifier)
+{
+    mpComm->SendControlSignal(Signal, rIdentifier);
+
+}
+inline Internals::ControlSignal CoSimIO::RecvControlSignal(std::string& rIdentifier)
+{
+    return mpComm->RecvControlSignal(rIdentifier);
+}
+
+
+inline void CoSimIO::RegisterAdvanceInTime(double (*pFuncPtr)(double))
+{
+    mpAdvInTime = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterInitializeSolutionStep(void (*pFuncPtr)())
+{
+    mpInitSolStep = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterSolveSolutionStep(void (*pFuncPtr)())
+{
+    mpSolSolStep = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterFinalizeSolutionStep(void (*pFuncPtr)())
+{
+    mpFinSolStep = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterImportGeometry(void (*pFuncPtr)(const std::string&))
+{
+    mpImportGeom = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterExportGeometry(void (*pFuncPtr)(const std::string&))
+{
+    mpExportGeom = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterImportMesh(void (*pFuncPtr)(const std::string&))
+{
+    mpImportMesh = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterExportMesh(void (*pFuncPtr)(const std::string&))
+{
+    mpExportMesh = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterImportData(void (*pFuncPtr)(const std::string&))
+{
+    mpImportData = pFuncPtr;
+}
+
+inline void CoSimIO::RegisterExportData(void (*pFuncPtr)(const std::string&))
+{
+    mpExportData = pFuncPtr;
+}
+
+
+inline void CoSimIO::Run()
```

Comment:
now I managed to completely hide this inside the API
The client only has to register the functions